File: dnase_enrichments.py
import glob
import numpy
import sys

gwasdir = "/mnt/lustre/home/cusanovich/500HT/MinalSherlock"
gwass = glob.glob(gwasdir + '/*sherlock*')
sensitivity = open("/mnt/lustre/home/cusanovich/500HT/Imputed1415/hutt.imputed.dnase.bed")
sensitivitydic = {}
print "Building sensitivity dictionary..."
for row in sensitivity:
	if 'chrm' in row:
		tissues = row.strip().split()
		continue
	rower = row.strip().split()
	sensitivitydic[rower[3]] = rower

sensitivity.close()
for gwas in gwass:
	currgwas = open(gwas,'r')
	currpheno = gwas.split('/')[-1].split('.')[0]
	outfile = open(gwasdir + '/' + currpheno + '.enrichment.txt.','w')
	print currpheno
	sys.stdout.flush()
	gwasdic = {}
	print "Building GWAS results dictionary..."
	sys.stdout.flush()
	for line in currgwas:
		liner = line.strip().split()
		gwasdic[liner[0]] = float(liner[1])
	currgwas.close()
	tiles = [0.025,0.05,0.1,0.2,0.4,0.8,1.6,3.2,6.4,12.8,25.6,50]
	tiled = [numpy.percentile(gwasdic.values(),x) for x in tiles]
	tiled.append(1.01)
	currsensdic = {}
	currtots = [0]*len(tiled)
	for tissue in tissues[4:]:
		currsensdic[tissue] = [0]*len(tiled)
	print "Calculating enrichments..."
	sys.stdout.flush()
	for snp in gwasdic.keys():
		snpp = gwasdic[snp]
		olaps = sensitivitydic[snp]
		for quant in range(len(tiled)):
			if snpp < tiled[quant]:
				currtots[quant] += 1
			for tissue in currsensdic.keys():
				if olaps[tissues.index(tissue)] != '0':
					currsensdic[tissue][quant] += 1
	print "Writing results..."
	sys.stdout.flush()
	print >> outfile, 'Tissue\t' + '\t'.join([str(x) for x in tiled])
	for tissue in currsensdic.keys():
		baseline = currsensdic[tissue][-1]/currtots[-1]
		print >> outfile, tissue + '\t' + '\t'.join([str((currsensdic[tissue][x]/currtots[x])/baseline) for x in range(len(currsensdic[tissue]))])
	outfile.close()

File: dnase_enrichments.R
print("Starting the engines...")
dnase = read.table("/mnt/lustre/home/cusanovich/500HT/Imputed1415/hutt.imputed.dnase2.bed", header=T)
gwass = list.files("/mnt/lustre/home/cusanovich/500HT/MinalSherlock", pattern="sherlock")
thresholds = c(0.00025,0.0005,0.001,0.002,0.004,0.008,0.016,0.032,0.064,0.128,0.256,0.5)
coloring = c("indianred", "darkred", "red", "dodgerblue2",
             "aquamarine", "cadetblue1", "mediumseagreen", "darkgreen",
             "yellowgreen", "mediumorchid3", "violetred2", "darkmagenta",
             "orange3", "coral4", "orangered2", "gold", "yellow3")
celltype = read.table("/mnt/lustre/home/cusanovich/500HT/stam_cells.txt")
cellcolors = coloring[unlist(as.factor(celltype[,2]))]
colorind = match(colnames(dnase)[5:length(colnames(dnase))],celltype[,1])
cellcolor = cellcolors[colorind]
badcells = which(is.na(cellcolor))
cellcolored = cellcolor[-badcells]
dnased = dnase[,-(badcells+4)]
celltype = NA
dnase = NA
pdf("/mnt/lustre/home/cusanovich/tissue_enrichments.pdf")
for(i in 1:length(gwass)){
	gwas = read.table(paste0("/mnt/lustre/home/cusanovich/500HT/MinalSherlock/",gwass[i]))
	currpheno = strsplit(gwass[i],"[.]")[[1]][1]
	print(currpheno)
	currind = match(gwas[,1],dnased[,4])
	currdnase = dnased[currind,5:dim(dnased)[2]]
	noners = which(is.na(currdnase[,1]))
	currdnase[noners,] = rep(0,dim(currdnase)[2])
	baselines = colSums(currdnase>0)/dim(currdnase)[1]
	currthresholds = quantile(gwas[,2],thresholds)
	currenrichments = matrix(NA,dim(currdnase)[2],length(currthresholds))
	currps = rep(1,dim(currdnase)[2])
	print("Calculating enrichments...")
	for(j in 1:dim(currdnase)[2]){
		for(k in 1:length(currthresholds)){
			currenrichments[j,k] = length(which(gwas[,2] < currthresholds[k] & currdnase[,j] > 0))/length(which(gwas[,2] < currthresholds[k]))/baselines[j]
			if(k == 1){
				olaps = length(which(gwas[,2] < currthresholds[k] & currdnase[,j] > 0))
				dnasers = length(which(currdnase[,j] > 0))
				nondnasers = dim(currdnase)[1] - dnasers
				chancers = length(which(gwas[,2] < currthresholds[k]))
				currps[j] = phyper((olaps - 1),dnasers,nondnasers,chancers, lower.tail=FALSE)
			}
		}
	}
	currqs = p.adjust(currps,method="BH")
	currenrichments = cbind(currenrichments,rep(1,times=dim(currdnase)[2]))
	colnames(currenrichments) = c(currthresholds,1)
	rownames(currenrichments) = colnames(currdnase)
	#write.table(currenrichments,paste0("/mnt/lustre/home/cusanovich/500HT/MinalSherlock/",currpheno,".enrichmenttable.txt"),quote=F,sep="\t")
	print("Plotting enrichments...")
	plot(-log10(as.numeric(colnames(currenrichments))),currenrichments[1,],type="l",ylim=c(0,max(currenrichments)),main=currpheno,col="gray")
	for(l in 2:dim(currenrichments)[1]){
		lines(-log10(as.numeric(colnames(currenrichments))),currenrichments[l,],col="gray")
	}
	for(m in 1:dim(currenrichments)[1]){
		if(currqs[m] < 0.2){
                	lines(-log10(as.numeric(colnames(currenrichments))),currenrichments[m,],col=cellcolored[m])
		}
        }
	currenrichments = cbind(currqs,currenrichments)
	colnames(currenrichments)[1] = "Qvals"
        currenrichments = cbind(currps,currenrichments)
        colnames(currenrichments)[1] = "Pvals"
	write.table(currenrichments,paste0("/mnt/lustre/home/cusanovich/500HT/MinalSherlock/",currpheno,".enrichmenttable.txt"),quote=F,sep="\t")
}
dev.off()
